src/MultiCollateralVault.sol

```diff
@@ -0,0 +1,275 @@
+    function _freeCollateral(address user, IERC20 token) internal view returns(uint256) {
+        uint256 userDebt =  debt[user];
+        IERC20[] memory _collaterals = Collaterals;
+        uint256 _length = _collaterals.length;
+        uint256 deposit_;
+        uint256 price_;
+        uint96 ratio_;
+        for (uint256 i = 0; i < _length; i++) {
+            if(_collaterals[i] != token && userDebt > 0) { 
+                deposit_ = deposits[user][_collaterals[i]]*10**_decimalsDifference(_collaterals[i]);
+                price_ = _price(_collaterals[i]);
+                ratio_ = collateralInfos[_collaterals[i]].ratio;
+                uint256 rdep = RMul.rmul(
+                    WMul.wmul(
+                    deposit_,
+                    price_), 
+                    ratio_);
+                if(userDebt > rdep){
+                    userDebt -= rdep;
+                } else {
+                    userDebt = 0;
+                }
+                ///unchecked{i++;}
+            }
+        }
+        if(userDebt > 0) {
+            return RMul.rmul( deposits[user][token], collateralInfos[token].ratio) - 
+                WDiv.wdiv(userDebt, _price(token))/10**_decimalsDifference(token);
+        } else{
+            return  deposits[user][token];
+        }
+    }
+
+    function _isCollateralized(address user) internal view returns(bool) {
+        (uint256 borrowingPower_, uint256 depositsInETH_) = _borrowingPower(user);
+        uint256 ltvRatio_ = WDiv.wdiv(borrowingPower_, debt[user]);
+        uint256 maxRatio_ = WDiv.wdiv(borrowingPower_, depositsInETH_);
+        bool collateralized;
+        if(ltvRatio_ <= maxRatio_){
+            collateralized = true;
+        } else {
+            collateralized = false;
+        }
+        return collateralized;
+    }
+/***
+-------------------------------------------------------------------------------------------------------------------------------------------------------------
+    User Functions
+-------------------------------------------------------------------------------------------------------------------------------------------------------------
+ */
+    
+    /// @notice allows a user to deposit WETH for use as collateral
+    /// @dev to ensure that a token is being accepted as collateral we check to ensure that a maximum loan to value ratio has been set since
+    /// that is the final step in adding a new token in collateral
+    function deposit(IERC20 token, uint256 amount) external {
+        require(collateralInfos[token].ratio != 0, "token not currently accepted as collateral");
+        deposits[msg.sender][token] += amount;
+        token.safeTransferFrom(msg.sender, address(this), amount);
+        emit Deposit(msg.sender, token, amount);
+    }
+
+    /// @notice allows a user to borrow WETH
+    function borrow(uint256 amount) external {
+        require(amount <= _remainingPower(msg.sender), "Insufficient collateral");
```

Comment:
You could instead `require(_isCollateralized(msg.sender))` _after_ adding the additional debt.

If adding the additional debt makes the user become undercollateralized, it reverts, and nothing happens.

Then you can remove `_remaininPower()`